server/src/main/java/org/elasticsearch/cluster/metadata/DataStreamAlias.java

```diff
@@ -221,38 +221,60 @@ public DataStreamAlias intersect(Predicate<String> filter) {
+        Set<String> mergedDataStreams = previous != null ? new HashSet<>(previous.getDataStreams()) : new HashSet<>();
 
         String writeDataStream = this.writeDataStream;
-        if (writeDataStream == null) {
-            if (other.getWriteDataStream() != null && mergedDataStreams.contains(other.getWriteDataStream())) {
-                writeDataStream = other.getWriteDataStream();
+        if (renamePattern != null && renameReplacement != null) {
+            this.dataStreams.stream().map(s -> s.replaceAll(renamePattern, renameReplacement)).forEach(mergedDataStreams::add);
+            if (writeDataStream != null) {
+                writeDataStream = writeDataStream.replaceAll(renamePattern, renameReplacement);
             }
+        } else {
+            mergedDataStreams.addAll(this.dataStreams);
         }
 
-        return new DataStreamAlias(this.name, List.copyOf(mergedDataStreams), writeDataStream, filter);
-    }
-
-    /**
-     * Returns a new instance with potentially renamed data stream names and write data stream name.
-     * If a data stream name matches with the provided rename pattern then it is renamed according
-     * to the provided rename replacement.
-     */
-    public DataStreamAlias renameDataStreams(String renamePattern, String renameReplacement) {
-        List<String> renamedDataStreams = this.dataStreams.stream()
-            .map(s -> s.replaceAll(renamePattern, renameReplacement))
-            .collect(Collectors.toList());
-        String writeDataStream = this.writeDataStream;
-        if (writeDataStream != null) {
-            writeDataStream = writeDataStream.replaceAll(renamePattern, renameReplacement);
+        if (previous != null) {
+            if (writeDataStream != null && previous.getWriteDataStream() != null) {
+                String previousWriteDataStream = previous.getWriteDataStream();
+                if (renamePattern != null && renameReplacement != null) {
+                    previousWriteDataStream = previousWriteDataStream.replaceAll(renamePattern, renameReplacement);
+                }
+                if (writeDataStream.equals(previousWriteDataStream) == false) {
+                    throw new IllegalArgumentException(
+                        "cannot merge alias ["
+                            + name
+                            + "], write data stream of this ["
+                            + writeDataStream
+                            + "] and write data stream of other ["
+                            + previous.getWriteDataStream()
+                            + "] are different"
+                    );
+                }
+            } else if (writeDataStream == null && previous.getWriteDataStream() != null) {
+                if (mergedDataStreams.contains(previous.getWriteDataStream())) {
+                    writeDataStream = previous.getWriteDataStream();
+                }
+                // else throw error?
```

Comment:
Does the case referenced by this comment need to be handled? I don't know how an alias would get into the state where it didn't have its write data stream, but it does seem like we shouldn't just fall through in that case.

Reply:
Also, do we need to handle the case where this instance has a write data stream and the `previous` instance does not have a write data stream?

Reply:
> Does the case referenced by this comment need to be handled? I don't know how an alias would get into the state where it didn't have its write data stream, but it does seem like we shouldn't just fall through in that case.

Previously we were also not handling this, this is why I added the comment here. I think we should handle it and throw an IllegalStateException if this happens.

> Also, do we need to handle the case where this instance has a write data stream and the previous instance does not have a write data stream?

We kind of handle this via an assertion in the constructor in `DataStreamAlias`. If we would handle this via a normal exception then I think it shouldn't be handled here but the mentioned constructor.

Reply:
I don't think it is possible to get into the state whether the write data stream of `previous` instance doesn't exist in `mergedDataStreams` set:
* Defining a data stream alias with a write data stream that does't exist in the list of data streams isn't possible. Mainly because when defining a write data stream then this is done via `is_write_index` boolean flag from the update aliases api and templates. Also there is the assertion in `DataStreamAlias` class' constructor. 
* The `mergedDataStreams` set is based on the data streams of both `this` and `previous` instances.

I've turned this into an assertion in the latest commit that I've pushed.